src/entities/postgres-storage/db-shelley.js

```diff
@@ -44,7 +59,114 @@ class DBShelley extends DB {
     await this.getConn().query(sql)
   }
 
-  async storeTx(tx: ShelleyTxType,
+  async getAccountDbData(accountInputs: Array<AccountInputType>): Promise<{
+  }> {
+    const accountIds = _.map(accountInputs, 'account_id')
+    const sql = Q.sql.select()
+      .from(Q.sql.select()
+        .from(ACCOUNTS_TBL)
+        .where('account in ?', accountIds)
+        .order('account')
+        .order('block_num', false)
+        .order('tx_ordinal', false)
+        .order('spending_counter', false)
+        .distinct('account'), 't')
+      .order('block_num', false)
+      .order('tx_ordinal', false)
+      .order('spending_counter', false)
+      .toString()
+    this.logger.debug('getAccountDbData', sql)
+    const dbRes = await this.getConn().query(sql)
+    let result = {}
+    for (const row of dbRes.rows) {
+      result = {
+        ...result,
+        ...{
+          [row.account]: {
+            balance: parseInt(row.balance, 10),
+            counter: parseInt(row.spending_counter, 10),
+          },
+        },
+      }
+    }
+    return result
+  }
+
+  async storeAccountsChanges(tx: TxType): Promise<void> {
+    const accountInputs = tx.inputs.filter(inp => inp.type === ACCOUNT_INP_TYPE)
+    const accountOutputs = tx.outputs.filter(out => out.type === ACCOUNT_INP_TYPE)
+    const allAccountIdsAndValues = [
+      ...accountInputs,
+      ...(accountOutputs).map(inp => ({
+        account_id: inp.address,
+        value: inp.value,
+      })),
+    ]
+    if (_.isEmpty(allAccountIdsAndValues)) {
+      return
+    }
+
+    const accountStoredData = await this.getAccountDbData(allAccountIdsAndValues)
+
+    const accountChanges = {}
+    accountInputs.forEach(account => {
+      const { account_id, value } = account
+      const currentChange = accountChanges[account_id]
+      if (currentChange !== undefined) {
+        accountChanges[account_id] = {
+          value: currentChange.value - value,
+          counter: currentChange.counter + 1,
+        }
+      } else {
+        accountChanges[account_id] = { value: 0 - value, counter: 1 }
```

Comment:
This shouldn't happen though, right? Maybe we should log this or something at least?

Reply:
Ah, it might happen because in this section is does not take in account the  current state yet, it only accumulates the **"total diff to each account per this tx"** so, e.g. if there's only one input for one account - the total diff will be negative, but then it should be applied to a positive current state.

But overall I agree, maybe we need to do some monitorings in case we ever get a negative balance